samza-core/src/main/java/org/apache/samza/coordinator/metadatastore/CoordinatorStreamStore.java

```diff
@@ -0,0 +1,198 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.coordinator.metadatastore;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Sets;
+import com.google.common.primitives.UnsignedBytes;
+import java.util.Collections;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.TreeMap;
+import org.apache.samza.Partition;
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.Config;
+import org.apache.samza.coordinator.stream.messages.CoordinatorStreamMessage;
+import org.apache.samza.metadatastore.MetadataStore;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.apache.samza.serializers.JsonSerde;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.system.IncomingMessageEnvelope;
+import org.apache.samza.system.SystemAdmin;
+import org.apache.samza.system.SystemConsumer;
+import org.apache.samza.system.SystemFactory;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemProducer;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.system.SystemStreamPartitionIterator;
+import org.apache.samza.system.SystemStreamMetadata;
+import org.apache.samza.system.SystemStreamMetadata.SystemStreamPartitionMetadata;
+import org.apache.samza.system.OutgoingMessageEnvelope;
+import org.apache.samza.util.CoordinatorStreamUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of the {@link MetadataStore} interface where the metadata of the Samza job is stored in coordinator stream.
+ *
+ * This class is thread safe.
+ */
+public class CoordinatorStreamStore implements MetadataStore {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CoordinatorStreamStore.class);
+  private static final String SOURCE = "SamzaContainer";
+
+  private final Config config;
+  private final SystemStream coordinatorSystemStream;
+  private final SystemStreamPartition coordinatorSystemStreamPartition;
+  private final SystemProducer systemProducer;
+  private final SystemConsumer systemConsumer;
+  private final SystemAdmin systemAdmin;
+  private final Object bootstrapLock = new Object();
```

Comment:
Move next to bootstrappedMessages to clarify what this is guarding.

Reply:
Done.